### PR TITLE
add connected shape based interpolation

### PR DESCRIPTION
**Pull request by dyollb**
Add first impl. of "connected shape-based interpolation", a slight modification of shape based interpolation, which aligns objects in top/bottom slices for interpolation and projects result back

### CODE REVIEW

File: Core/ConnectedShapeBasedInterpolation.h [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+namespace iseg {
+
+class ConnectedShapeBasedInterpolation
+{
+public:
+	using image_type = itk::Image<float, 2>;
+	using mask_type = itk::Image<unsigned char, 2>;
+	using image_stack_type = itk::SliceContiguousImage<float>;
+
+	std::vector<image_type::Pointer> interpolate(const image_type* slice1, const image_type* slice2, int number_of_slices)
```
Comment (sanderegg) [inline Core/ConnectedShapeBasedInterpolation.h]: aren't itk::image already shared pointers? why pass them as pointer?
can number of slice be < 0? I would pass an index (either unsigned int or size_t)

File: Core/ConnectedShapeBasedInterpolation.h [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+		int num_objects1 = 0;
+		int num_objects2 = 0;
+		auto objects1 = get_components(slice1, num_objects1);
+		auto objects2 = get_components(slice2, num_objects2);
+		if (num_objects1 != 1 || num_objects2 != 1)
+		{
+			throw std::runtime_error("Detected multiple foreground objects");
+		}
+
+		auto x1 = get_center_of_mass(objects1, num_objects1)[0];
```
Comment (sanderegg) [inline Core/ConnectedShapeBasedInterpolation.h]: is the use of const in front of auto not part of the conventions? would make sense here...

File: Core/ConnectedShapeBasedInterpolation.h [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+		std::vector<float*> slices;
+		slices.push_back(sdf1->GetPixelContainer()->GetImportPointer());
+		slices.push_back(sdf2->GetPixelContainer()->GetImportPointer());
+
+		auto container = image_stack_type::PixelContainer::New();
+		container->SetImportPointersForSlices(slices, size[0] * size[1], false);
+		image3d->SetPixelContainer(container);
+
+		std::vector<image_type::Pointer> interpolated_slices;
+		for (int i = 1; i <= number_of_slices; i++)
```
Comment (sanderegg) [inline Core/ConnectedShapeBasedInterpolation.h]: not sure if it's still the case due to possible compiler optimisations, but pre-incrementing needs less copying than post-incrementing
and again, are signed integer needed here?

File: Core/ConnectedShapeBasedInterpolation.h [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+		{
+			auto slice = interpolate(image3d, i / (number_of_slices + 1.0), translation);
+			interpolated_slices.push_back(slice);
+		}
+		return interpolated_slices;
+	}
+
+private:
+	/// find connected foreground regions (called "objects")
+	mask_type::Pointer get_components(const image_type* img, int& num_objects) const
```
Comment (sanderegg) [inline Core/ConnectedShapeBasedInterpolation.h]: if itk::image are already smart pointers (which I think they are)... pass as const reference

File: Core/ConnectedShapeBasedInterpolation.h [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+		auto id = itk::IdentityTransform<double, 3>::New();
+		auto origin = image3d->GetOrigin();
+		origin[2] = ratio_12;
+		auto size = image3d->GetBufferedRegion().GetSize();
+		size[2] = 1;
+
+		using image_3d_type = itk::Image<float, 3>;
+		auto resample_filter = itk::ResampleImageFilter<image_stack_type, image_3d_type>::New();
+		resample_filter->SetTransform(id.GetPointer());
+		resample_filter->SetInput(image3d);
```
Comment (sanderegg) [inline Core/ConnectedShapeBasedInterpolation.h]: I think all these filters also have a set multi threaded function (do not remember the exact syntax which can be of intereset performance wise)

File: iSeg/InterpolationWidget.cpp [hunk 6]
```diff
@@ -205,6 +208,7 @@ void InterpolationWidget::startslice_pressed()
 void InterpolationWidget::execute()
 {
 	unsigned short batchstride = (unsigned short)sb_batchstride->value();
+	bool const connected = cb_connectedshapebased->isVisible() && cb_connectedshapebased->isChecked();
```
Comment (sanderegg) [inline iSeg/InterpolationWidget.cpp]: hmm bool const, not const bool?

File: Core/ConnectedShapeBasedInterpolation.h (hunk 2 above)
Comment (dyollb) [reply Core/ConnectedShapeBasedInterpolation.h]: ok, will fix this

File: Core/ConnectedShapeBasedInterpolation.h (hunk 3 above)
Comment (dyollb) [reply Core/ConnectedShapeBasedInterpolation.h]: sure

File: Core/ConnectedShapeBasedInterpolation.h (hunk 4 above)
Comment (dyollb) [reply Core/ConnectedShapeBasedInterpolation.h]: you mean return? I think I need to return a ::Pointer to keep it alive.

File: Core/ConnectedShapeBasedInterpolation.h (hunk 5 above)
Comment (dyollb) [reply Core/ConnectedShapeBasedInterpolation.h]: I will double-check, however, I thought they follow global options, e.g. based on environment variables, etc.

File: iSeg/InterpolationWidget.cpp (hunk 6 above)
Comment (dyollb) [reply iSeg/InterpolationWidget.cpp]: you don't like the style?

File: Core/ConnectedShapeBasedInterpolation.h (hunk 4 above)
Comment (sanderegg) [reply Core/ConnectedShapeBasedInterpolation.h]: no I meant the arguments. the return value is clear

File: Core/ConnectedShapeBasedInterpolation.h (hunk 5 above)
Comment (sanderegg) [reply Core/ConnectedShapeBasedInterpolation.h]: that may also be.

File: iSeg/InterpolationWidget.cpp (hunk 6 above)
Comment (sanderegg) [reply iSeg/InterpolationWidget.cpp]: nope;) but nevermind...

File: Core/ConnectedShapeBasedInterpolation.h [hunk 7]
```diff
@@ -130,7 +208,7 @@ class ConnectedShapeBasedInterpolation
 		}
 
 		std::vector<itk::Point<double, 2>> pts(num_objects);
-		for (int id=0; id<num_objects; ++id)
+		for (int id = 0; id < num_objects; ++id)
```
Comment (sanderegg) [inline Core/ConnectedShapeBasedInterpolation.h]: why everywhere are int used instead of unsigned int? these are indices aren't they?

**Pull request merged**